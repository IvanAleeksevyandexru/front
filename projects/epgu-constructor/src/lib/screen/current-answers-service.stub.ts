import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';

@Injectable()
export class CurrentAnswersServiceStub {
  _isValid = new BehaviorSubject<boolean>(null);

  _state: string | object;

  get state(): string | object {
    return this._state || '{"foo": "bar"}';
  }

  set state(value) {
    this._state = value || {};
  }

  get isValid(): boolean {
    return this._isValid.getValue();
  }

  set isValid(val: boolean) {
    this._isValid.next(val);
  }

  get isValid$(): Observable<boolean> {
    return this._isValid.asObservable();
  }
}
