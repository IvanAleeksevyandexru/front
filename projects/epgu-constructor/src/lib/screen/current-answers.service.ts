import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';

@Injectable()
export class CurrentAnswersService {
  private _state: string | object;

  private _isValid = new BehaviorSubject<boolean>(null);

  get state(): string | object {
    return this._state;
  }

  set state(val: string | object) {
    this._state = val;
  }

  get isValid(): boolean {
    return this._isValid.getValue();
  }

  set isValid(val: boolean) {
    this._isValid.next(val);
  }

  get isValid$(): Observable<boolean> {
    return this._isValid.asObservable();
  }
}
