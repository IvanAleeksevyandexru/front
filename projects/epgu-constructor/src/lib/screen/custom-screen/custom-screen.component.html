<epgu-cf-ui-screen-container
  [show-nav]="screenService.showNav$ | async"
  [class.screen-container-mt]="!(screenService.showNav$ | async)"
  [buttons$]="screenService.buttons$"
>
  <div class="screen-body" *ngIf="data$ | async as data">
    <epgu-constructor-page-name class="mb-0">{{ screenService.header$ | async }} </epgu-constructor-page-name>

    <epgu-cf-ui-constructor-screen-pad
      class="body"
      [ngClass]="{'disabled-white-bg': disableWhiteBackground}"
    >
      <epgu-constructor-components-list
        [components]="$any(data?.components)"
        [errors]="screenService.componentErrors$ | async"
        (changes)="changeComponentsList($event)"
      >
      </epgu-constructor-components-list>
    </epgu-cf-ui-constructor-screen-pad>

    <epgu-constructor-label
      [isTextHelper]="helperText?.attrs?.isTextHelper"
      [tips]="helperText?.attrs?.labelHint"
      [label]="helperText?.label"
      [clarifications]="helperText?.attrs?.clarifications"
      [largeFontSize]="true"
      class="mt-24"
      *ngIf="helperText"
    ></epgu-constructor-label>
  </div>

  <div class="screen-footer">
    <epgu-constructor-screen-buttons
      class="screen-buttons"
      *ngIf="(screenService.buttons$ | async).length"
      [screenButtons]="screenService.buttons$ | async"
      [isLoading]="screenService.isLoading$ | async"
      [disabled]="!isValid"
    ></epgu-constructor-screen-buttons>
  </div>
</epgu-cf-ui-screen-container>
