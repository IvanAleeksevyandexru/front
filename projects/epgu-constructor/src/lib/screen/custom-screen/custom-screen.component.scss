@import '~@epgu/epgu-constructor-ui-kit/src/styles/core';

:host {
  @include flex-container();
}

.body {
  display: flex;
  text-align: left;
  @include tablets-down {
    flex-grow: 1;
  }
  > * {
    display: flex;
    width: 100%;
    flex-direction: column;
  }
}

@include screen-buttons-margin;
