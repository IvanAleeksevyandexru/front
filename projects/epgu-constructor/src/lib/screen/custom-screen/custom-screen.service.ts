import { Injectable } from '@angular/core';
import { CustomComponentOutputData } from '../../component/custom-screen/components-list.types';
import { DatesToolsService } from '@epgu/epgu-constructor-ui-kit';
import { NavigationPayload } from '../../form-player/form-player.types';

@Injectable()
export class CustomScreenService {
  constructor(private datesToolsService: DatesToolsService) {}

  /**
   * Форматируем данные перед отправкой
   * @param changes - данные на отправку
   */
  getFormattedData(changes: CustomComponentOutputData): NavigationPayload {
    return this.getPrepareResponseData(changes);
  }

  /**
   * Подготавливаем данные для ответа
   * @param data - данные для пребразования
   * @private
   */
  private getPrepareResponseData(data: CustomComponentOutputData = {}): NavigationPayload {
    return Object.keys(data).reduce<NavigationPayload>((acc, key) => {
      let value = '';
      const dataValue = data[key].value;

      if (typeof dataValue === 'object') {
        if (this.datesToolsService.isValid(dataValue)) {
          value = this.datesToolsService.format(dataValue);
        } else {
          value = dataValue ? JSON.stringify(dataValue ) : null;
        }
      } else {
        value = dataValue;
      }

      const payload = {
        visited: true,
        value,
        disabled: data[key].disabled,
      };

      return {
        ...acc,
        [key]: payload,
      };
    }, {});
  }
}
