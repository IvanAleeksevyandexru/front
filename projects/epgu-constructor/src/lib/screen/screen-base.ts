import { Injector } from '@angular/core';
import { EventBusService } from '@epgu/epgu-constructor-ui-kit';
import { NavigationService } from '../core/services/navigation/navigation.service';
import { UnsubscribeService } from '@epgu/epgu-constructor-ui-kit';
import { CurrentAnswersService } from './current-answers.service';
import { ScreenService } from './screen.service';

// Желательно больше не наследоваться потому что это может усложнить логику.
export abstract class ScreenBase {
  public navigationService: NavigationService;
  public ngUnsubscribe$: UnsubscribeService;
  public screenService: ScreenService;
  public currentAnswersService: CurrentAnswersService;
  public eventBusService: EventBusService;

  protected constructor(public injector: Injector) {
    this.navigationService = this.injector.get(NavigationService);
    this.ngUnsubscribe$ = this.injector.get(UnsubscribeService);
    this.screenService = this.injector.get(ScreenService);
    this.currentAnswersService = this.injector.get(CurrentAnswersService);
    this.eventBusService = this.injector.get(EventBusService);
  }
}
