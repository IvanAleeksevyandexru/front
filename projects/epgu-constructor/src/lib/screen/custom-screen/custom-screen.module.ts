import { NgModule } from '@angular/core';
import { ComponentsListModule } from '../../component/custom-screen/components-list.module';
import { BaseModule } from '../../shared/base.module';
import { BaseComponentsModule } from '../../shared/components/base-components/base-components.module';
import { ScreenContainerModule } from '@epgu/epgu-constructor-ui-kit';
import { ScreenPadModule } from '@epgu/epgu-constructor-ui-kit';
import { CustomScreenComponent } from './custom-screen.component';
import { CustomScreenService } from './custom-screen.service';
import { ScreenButtonsModule } from '../../shared/components/screen-buttons/screen-buttons.module';

@NgModule({
  declarations: [CustomScreenComponent],
  exports: [CustomScreenComponent],
  imports: [
    BaseModule,
    ComponentsListModule,
    BaseComponentsModule,
    ScreenContainerModule,
    ScreenPadModule,
    ScreenButtonsModule,
  ],
  providers: [
    CustomScreenService,
  ]
})
export class CustomScreenModule {}
