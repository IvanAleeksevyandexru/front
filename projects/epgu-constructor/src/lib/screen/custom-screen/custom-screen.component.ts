import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  Injector,
  OnInit,
} from '@angular/core';
import { UnsubscribeService } from '@epgu/epgu-constructor-ui-kit';

import {
  CustomComponent,
  CustomComponentOutputData,
  CustomComponentValidationConditions,
  CustomScreenComponentTypes,
} from '../../component/custom-screen/components-list.types';
import { NavigationPayload } from '../../form-player/form-player.types';
import { ScreenBase } from '../screen-base';
import { CustomScreenService } from './custom-screen.service';

@Component({
  selector: 'epgu-constructor-custom-screen',
  templateUrl: './custom-screen.component.html',
  styleUrls: ['./custom-screen.component.scss'],
  providers: [UnsubscribeService],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CustomScreenComponent extends ScreenBase implements OnInit {
  dataToSend: NavigationPayload;
  isValid: boolean;
  helperText: CustomComponent;

  constructor(
    public injector: Injector,
    private customScreenService: CustomScreenService,
    private cdr: ChangeDetectorRef,
  ) {
    super(injector);
  }

  ngOnInit(): void {
    this.setHelperText();
  }

  /**
   * Форматируем данные перед отправкой
   * @param changes - данные на отправку
   */
  changeComponentsList(changes: CustomComponentOutputData): void {
    const notAtLeastOne = Object.values(changes).filter(
      (item) => item.condition !== CustomComponentValidationConditions.atLeastOne,
    );
    const atLeastOne = Object.values(changes).filter(
      (item) => item.condition === CustomComponentValidationConditions.atLeastOne,
    );

    const notAtLeastOneExpression: boolean = notAtLeastOne.length
      ? notAtLeastOne.every((item) => item.isValid)
      : true;

    const atLeastOneExpression: boolean = atLeastOne.length
      ? atLeastOne.some((item) => item.value) && atLeastOne.every((item) => item.isValid)
      : true;

    this.isValid = notAtLeastOneExpression && atLeastOneExpression;
    this.dataToSend = this.customScreenService.getFormattedData(changes);
    this.currentAnswersService.isValid = this.isValid;
    this.currentAnswersService.state = this.dataToSend;
    this.cdr.detectChanges();
  }

  private setHelperText(): void {
    const components = this.screenService.display?.components;
    let helperTextIndex = -1;
    this.helperText = components?.find((component: CustomComponent, index: number) => {
      if (
        component.attrs.isTextHelper &&
        component.attrs.isBottomSlot &&
        component.type === CustomScreenComponentTypes.LabelSection
      ) {
        helperTextIndex = index;
        return true;
      }
      return false;
    }) as CustomComponent;
    if (helperTextIndex > -1) {
      components?.splice(helperTextIndex, 1);
    }
  }
}
