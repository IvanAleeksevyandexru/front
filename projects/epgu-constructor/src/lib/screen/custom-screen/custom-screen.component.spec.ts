import { ChangeDetectionStrategy, NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { MockComponent, MockModule } from 'ng-mocks';
import { ComponentsListComponent } from '../../component/custom-screen/components-list.component';
import {
  CustomComponentOutputData,
  CustomComponentValidationConditions,
} from '../../component/custom-screen/components-list.types';
import {
  LoggerService,
  ConfigService,
  ModalService,
  ModalServiceStub,
  ScreenPadComponent,
  DatesToolsService,
  EventBusService,
} from '@epgu/epgu-constructor-ui-kit';
import { NavigationService } from '../../core/services/navigation/navigation.service';
import { NavigationServiceStub } from '../../core/services/navigation/navigation.service.stub';
import { PageNameComponent } from '../../shared/components/base-components/page-name/page-name.component';
import { ScreenContainerComponent } from '@epgu/epgu-constructor-ui-kit';
import { CurrentAnswersService } from '../current-answers.service';
import { ScreenService } from '../screen.service';
import { ScreenServiceStub } from '../screen.service.stub';
import { CustomScreenComponent } from './custom-screen.component';
import { CustomScreenService } from './custom-screen.service';
import { UserInfoLoaderModule } from '../../shared/components/user-info-loader/user-info-loader.module';
import { BaseModule } from '../../shared/base.module';
import { ScreenButtonsModule } from '../../shared/components/screen-buttons/screen-buttons.module';
import { ActionService } from '../../shared/directives/action/action.service';
import { ActionServiceStub } from '../../shared/directives/action/action.service.stub';
import {
  ComponentDto,
  ActionType,
  DTOActionAction,
  ScreenTypes,
  DisplayDto
} from '@epgu/epgu-constructor-types';
import { EaisdoGroupCostService } from '../../shared/services/eaisdo-group-cost/eaisdo-group-cost.service';
import { CertificateEaisdoService } from '../../shared/services/certificate-eaisdo/certificate-eaisdo.service';
import { HttpClientModule } from '@angular/common/http';

describe('CustomScreenComponent', () => {
  let component: CustomScreenComponent;
  let fixture: ComponentFixture<CustomScreenComponent>;

  let screenService: ScreenServiceStub;
  let customScreenService: CustomScreenService;
  let datesToolsService: DatesToolsService;
  let components = [
    {
      id: 'bd01',
      type: 'LabelSection',
      label: 'Нажимая “Отправить заявление”, вы соглашаетесь с правилами предоставления услуги',
      attrs: {
        isTextHelper: true,
        isBottomSlot: true,
      },
      value: '',
      visited: false,
    },
    {
      id: 'bd1',
      type: 'LabelSection',
      label: 'Нажимая “Отправить заявление”, вы соглашаетесь с правилами предоставления услуги',
      attrs: {
        // @ts-ignore
        isTextHelper: true,
      },
      value: '',
      visited: false,
    },
  ] as ComponentDto[];

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        MockModule(UserInfoLoaderModule),
        BaseModule,
        ScreenButtonsModule,
        HttpClientModule,
      ],
      declarations: [
        CustomScreenComponent,
        MockComponent(PageNameComponent),
        MockComponent(ComponentsListComponent),
        MockComponent(ScreenPadComponent),
        MockComponent(ScreenContainerComponent),
      ],
      providers: [
        { provide: ScreenService, useClass: ScreenServiceStub },
        { provide: NavigationService, useClass: NavigationServiceStub },
        { provide: ActionService, useClass: ActionServiceStub },
        { provide: ModalService, useClass: ModalServiceStub },
        EventBusService,
        CurrentAnswersService,
        CertificateEaisdoService,
        CustomScreenService,
        DatesToolsService,
        ConfigService,
        LoggerService,
        EaisdoGroupCostService,
      ],
      schemas: [NO_ERRORS_SCHEMA],
    }).overrideComponent(CustomScreenComponent, {
      set: { changeDetection: ChangeDetectionStrategy.Default },
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CustomScreenComponent);
    component = fixture.componentInstance;
    screenService = (TestBed.inject(ScreenService) as unknown) as ScreenServiceStub;
    customScreenService = TestBed.inject(CustomScreenService);
    datesToolsService = TestBed.inject(DatesToolsService);
    screenService.buttons = [];
    screenService.display = {
      components,
      header: 'header1',
      id: 'id1',
      name: 'name1',
      type: ScreenTypes.CUSTOM,
      terminal: true,
    } as DisplayDto;

    component.ngOnInit();
    fixture.detectChanges();
  });

  describe('ngOnInit()', () => {
    it('should call setHelperText()', () => {
      const spy = jest.spyOn<any, string>(component, 'setHelperText');
      component.ngOnInit();
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('setHelperText()', () => {
    it('should mutate initial components list length, if there is LabelSection with specific attrs among them', () => {
      expect(screenService.display.components.length).toBe(1);
    });
  });

  describe('dataToSend property', () => {
    it('should be undefined by default', () => {
      expect(component.dataToSend).toBeUndefined();
    });
  });

  describe('isValid property', () => {
    it('should be undefined by default', () => {
      expect(component.isValid).toBeUndefined();
    });
  });

  describe('getFormattedData() method', () => {
    it('without data', () => {
      const date = new Date();
      const changes: CustomComponentOutputData = {
        any: {
          value: datesToolsService.format(date),
          valid: true,
        },
      };
      const expectedResult = {
        any: { visited: true, disabled: undefined, value: datesToolsService.format(date) },
      };

      const result = customScreenService.getFormattedData(changes);

      expect(result).toEqual(expectedResult);
    });
  });

  describe('changeComponentList() method', () => {
    // @todo. fix type
    const expectedDataSend: CustomComponentOutputData = ({
      any: {
        value: 'any',
        visited: true,
        disabled: undefined,
      },
    } as unknown) as CustomComponentOutputData;

    it('condition === atLeastOne', () => {
      // @todo. fix type
      const changes: CustomComponentOutputData = ({
        any: {
          condition: CustomComponentValidationConditions.atLeastOne,
          isValid: true,
          value: 'any',
        },
      } as unknown) as CustomComponentOutputData;

      component.changeComponentsList(changes);

      expect(component.isValid).toBeTruthy();
      expect(component.dataToSend).toEqual(expectedDataSend);
    });

    it('condition !== atLeastOne', () => {
      // @todo. fix type
      const changes: CustomComponentOutputData = ({
        any: {
          condition: 'any',
          isValid: false,
          value: 'any',
        },
      } as unknown) as CustomComponentOutputData;

      component.changeComponentsList(changes);

      expect(component.isValid).toBeFalsy();
      expect(component.dataToSend).toEqual(expectedDataSend);
    });
  });

  describe('epgu-cf-ui-screen-container', () => {
    const selector = 'epgu-cf-ui-screen-container';

    it('should be rendered', () => {
      const debugEl = fixture.debugElement.query(By.css(selector));

      expect(debugEl).toBeTruthy();
    });

    it('showNav property should be TRUE if screenService.showNav is TRUE, otherwise should be FALSE', () => {
      const debugEl = fixture.debugElement.query(By.css(selector));

      expect(debugEl.componentInstance.showNav).toBeFalsy();

      screenService.showNav = false;
      fixture.detectChanges();

      expect(debugEl.componentInstance.showNav).toBeFalsy();

      screenService.showNav = true;
      fixture.detectChanges();

      expect(debugEl.componentInstance.showNav).toBeTruthy();
    });
  });

  it('should render epgu-constructor-page-name', () => {
    const selector = 'epgu-cf-ui-screen-container epgu-constructor-page-name';

    const debugEl = fixture.debugElement.query(By.css(selector));

    expect(debugEl).toBeTruthy();
  });

  it('should render epgu-cf-ui-constructor-screen-pad', () => {
    const selector = 'epgu-cf-ui-screen-container epgu-cf-ui-constructor-screen-pad';

    const debugEl = fixture.debugElement.query(By.css(selector));

    expect(debugEl).toBeTruthy();
  });

  it('should render epgu-constructor-components-list', () => {
    const selector =
      'epgu-cf-ui-screen-container epgu-cf-ui-constructor-screen-pad epgu-constructor-components-list';

    const debugEl = fixture.debugElement.query(By.css(selector));

    expect(debugEl).toBeTruthy();
    expect(debugEl.componentInstance.errors).toBeNull();

    const components: ComponentDto[] = [
      {
        attrs: {},
        id: 'id1',
        type: 'type1',
      },
    ];
    screenService.display = {
      components,
      header: 'header1',
      id: 'id1',
      name: 'name1',
      type: ScreenTypes.CUSTOM,
      terminal: true,
    };

    const componentErrors = (screenService.componentErrors = {
      foo: 'bar',
    });

    fixture.detectChanges();

    expect(debugEl.componentInstance.components).toBe(components);
    expect(debugEl.componentInstance.errors).toBe(componentErrors);
  });

  it('should call changeComponentsList() on epgu-constructor-components-list changes() event', () => {
    const selector =
      'epgu-cf-ui-screen-container epgu-cf-ui-constructor-screen-pad epgu-constructor-components-list';

    const debugEl = fixture.debugElement.query(By.css(selector));

    const date = new Date();
    const changesData: CustomComponentOutputData = {
      any: {
        value: date.toISOString(),
        valid: true,
      },
    };

    const changeComponentsListSpy = jest.spyOn(component, 'changeComponentsList');

    debugEl.triggerEventHandler('changes', changesData);

    expect(changeComponentsListSpy).toBeCalledTimes(1);
    expect(changeComponentsListSpy).toBeCalledWith(changesData);
  });

  describe('Submit button', () => {
    const selector = 'epgu-cf-ui-screen-container lib-button';

    it('should be rendered if has buttons', () => {
      let debugEl = fixture.debugElement.query(By.css(selector));

      expect(debugEl).toBeNull();

      screenService.buttons = [
        {
          label: 'any',
          action: DTOActionAction.getNextStep,
          type: ActionType.nextStep,
        },
      ];
      fixture.detectChanges();

      debugEl = fixture.debugElement.query(By.css(selector));

      expect(debugEl).toBeTruthy();
    });

    it('disabled property should be TRUE if screenService.isLoading is TRUE or isValid is FALSE', () => {
      screenService.buttons = [
        {
          label: 'any',
          action: DTOActionAction.getNextStep,
          type: ActionType.nextStep,
        },
      ];

      screenService.isLoadingSubject.next(true);
      component.isValid = false;

      fixture.detectChanges();

      const debugEl = fixture.debugElement.query(By.css(selector));

      expect(debugEl.componentInstance.disabled).toBeTruthy();

      screenService.isLoadingSubject.next(false);
      fixture.detectChanges();

      expect(debugEl.componentInstance.disabled).toBeTruthy();

      component.isValid = true;
      fixture.detectChanges();

      expect(debugEl.componentInstance.disabled).toBeFalsy();
    });
  });
});
