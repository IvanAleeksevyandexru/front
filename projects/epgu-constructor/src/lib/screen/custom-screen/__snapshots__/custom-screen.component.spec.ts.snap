// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomScreenComponent check snapshot 1`] = `
<epgu-constructor-custom-screen
  currentAnswersService={[Function CurrentAnswersService]}
  customScreenService={[Function CustomScreenService]}
  eventBusService={[Function EventBusService]}
  helperText="undefined"
  injector={[Function Injector_]}
  navigationService={[Function NavigationServiceStub]}
  ngUnsubscribe$={[Function UnsubscribeService]}
  screenService={[Function ScreenServiceStub]}
>
  <epgu-cf-ui-screen-container
    class="screen-container-mt"
  >
    <div
      class="screen-body"
    >
      <epgu-constructor-page-name>
         
        <epgu-constructor-user-info-loader />
      </epgu-constructor-page-name>
      <epgu-cf-ui-constructor-screen-pad
        class="body"
      >
        <epgu-constructor-components-list />
      </epgu-cf-ui-constructor-screen-pad>
      
    </div>
    <div
      class="screen-footer"
    >
      
    </div>
  </epgu-cf-ui-screen-container>
</epgu-constructor-custom-screen>
`;
